### DISCUSSION TITLE
LAB#C codec library:run makefile fail and run L1 cosim fail

### DISCUSSION THREAD

**Discussion by s095339** [bol-edu/HLS-SOC-Discussions#46] [Q&A] [answered]
Hello, we are 張耀明 and 陳佳詳, we are students from nthu.
We have been doing Lab-C, Codec vitis library, using the U50_04.
 We encountered Two problems：

1. Fail to run jpegdec L1-API cosim.
We did the following operations before running cosim
first I configured the environment
```
cd /Vitis_Libraries/codec/L1/tests/jpegdec
source "/opt/Xilinx/Vitis_HLS/2022.1/settings64.sh"
source "/opt/Xilinx/Vitis/2022.1/settings64.sh"
source "/opt/xilinx/xrt/setup.sh"
export PLATFORM_REPO_PATHS="/opt/xilinx/platforms/xilinx_u50_gen3x16_xdma_5_202210_1/xilinx_u50_gen3x16_xdma_5_202210_1.xpfm"
export DEVICE= xilinx_u50_gen3x16_xdma_5_202210_1
export TARGET=sw_emu
```
Then I ran the Csim Csynth Cosim
```
#Csim
make run DEVICE=xilinx_u50_gen3x16_xdma_5_202210_1  CSIM=1
#Csyn
make run DEVICE=xilinx_u50_gen3x16_xdma_5_202210_1  CSYNTH=1
#Cosim
make run DEVICE=xilinx_u50_gen3x16_xdma_5_202210_1  COSIM=1
```
the Csim and Csyn can work with no error message, but we failed to run Cosim.
Here is the log file generated after running the Cosim.

[vitis_hls.log](https://github.com/bol-edu/HLS-Discussions/files/11296164/vitis_hls.log)

2. Fail to compile jpegdec L2-api host program
```
cd /Vitis_Libraries/codec/L2/demos/jpegDec
source "/opt/Xilinx/Vitis_HLS/2022.1/settings64.sh"
source "/opt/Xilinx/Vitis/2022.1/settings64.sh"
make clean
source "/opt/xilinx/xrt/setup.sh"
export PLATFORM_REPO_PATHS="/opt/xilinx/platforms/"
export TARGET=sw_emu
export DEVICE= xilinx_u50_gen3x16_xdma_5_202210_1
make host
```

![image](https://user-images.githubusercontent.com/78518884/233655724-67fc5331-a4b1-4d95-9c51-dfd0dc81b2dd.png)
it says "cannot find  Ip_floating_point_v7_1_bitacc_cmodel" .

I comment it out in the makefile so that the "make host" can be run with no error.
![image](https://user-images.githubusercontent.com/78518884/233657096-c51fed06-e17c-4211-84f1-8e36cfc15995.png)

Then I run
`make run`
![image](https://user-images.githubusercontent.com/78518884/233658067-e3fc8580-2763-4859-862a-cde54b65253e.png)

 Everything looks fine, the .yuv can be successfully generated.
![image](https://user-images.githubusercontent.com/78518884/233657246-74eaa088-8647-473e-ab95-547604f62725.png)
But I am not sure if the result is generated with no problem.
the .yuv image:
![image](https://user-images.githubusercontent.com/78518884/233656905-30fa73e0-38d1-40f2-8461-db0077c74100.png)

**Comment by boledulab** [comment, marked as answer]
Hi,
The first problem was known offical issue, https://support.xilinx.com/s/article/Use-of-gmp-h-for-Co-simulation
The problem was fixed by (only) adminstrator.
change in /opt/Xilinx/Vitis_HLS/2022.1/include/mpfr.h
```
//#ifndef __GMP_H__
#include <gmp.h>
#define __gmp_const const
//#endif
```
[co-sim log](https://github.com/bol-edu/HLS-Discussions/files/11301037/co-sim.log)
offical doc: https://docs.xilinx.com/r/jpiLUWqYENGWAj4YktN~YQ/~8rb6JHOcKsEZ2KdYnULmA

**Reply by s095339** [reply]
Thanks, I can run it successfully

**Comment by boledulab** [comment]
For second problem, I successfully run "make host" && "make run" without any change.
Maybe it was fixed by corrected gmp.h issue.

[make host && make run log](https://github.com/bol-edu/HLS-Discussions/files/11301085/make-host_make-run.log)

**Reply by s095339** [reply]
Sorry, we still can't run due to the same problem.

**Reply by kevinjantw** [reply]
I passed "make host" && "make run" because I used your test code already removing "-lIp_floating_point_v7_1_bitacc_cmodel". 
https://github.com/bol-edu/HLS-Discussions/files/11301085/make-host_make-run.log
In order to resolve second problem you need to setup library from Vivado IDE, please refer the doc chapter 5. 
https://www.xilinx.com/content/dam/xilinx/support/documents/ip_documentation/floating_point/v7_1/pg060-floating-point.pdf  

From chapter 5@Floating-Point Operator v7.1 LogiCORE IP Product Guide:
The Xilinx® LogiCORE ™ IP Floating-Point Operator core bit accurate C model is a
self-contained, linkable, shared library that models the functionality of this core with finite
precision arithmetic. This model provides a bit accurate representation of the various
modes of the `Floating-Point Operator v7.1 core`, and it is suitable for inclusion in a larger
framework for system-level simulation or core-specific verification.  

`The C model is an optional output of the Vivado® Design Suite`. For information about
generating IP source outputs, see Vivado Design Suite User Guide: Using the Vivado IDE
(UG893)

**Reply by s095339** [reply]
Sorry for the delay,
I've check the doc chapter 5.
In the doc, it says:
![image](https://user-images.githubusercontent.com/78518884/235610540-7bce36b4-fe41-450d-8b18-7d86b882f194.png)
But I can't find where these .so files are located

**Comment by kevinjantw** [comment]
Hi,
According 'The C model is an optional output of the Vivado® Design Suite', maybe you can generate C model (*.so) from Vivado. 
You can explore Xilinx LogiCORE IP design flow from Xilinx community posts.
